udev: introduce TxQueueLength= setting

OnkelUlla opened the pull request:
Enable udev to set the transmit queue length of a device via a new directive to
be used in link files.

yuwata commented:
Hmm, instead of .link file, maybe we should add this in .network file?
Not sure, is the setting is related to the recently added settings by 0a9fb9bad8b3d245ae4dc27e5fd2675672ddd9ec?

poettering commented on man/systemd.link.xml:
```diff
@@ -409,6 +409,14 @@
+        <listitem>
+          <para>Specifies the transmit queue length of the device in number of packets. An unsigned integer
+          in the range 0..4294967295. When unset (equivalent to explicitely setting it to 0), the kernel's
+          default will be used.</para>
```
don't make "0" magic here. 

Our usual way to reset values to their defaults is to assign an empty string. do that here too. i.e. refuse "0", but set the internal value to 0 if the empty string is assigned.

i.e. the fact that 0 means "kernel default" should be an internal detail, and not exposed to the user. and we should stick to the same logic everywhere for resetting values to their defaults: empty string.

OnkelUlla replied:
Thanks for clarifying that to me: I studied diverse related pull requests afterwards and saw that you did not explain this for the first time... unfortunately I fear that this hasn't been the last time either.

And it was also good to look after things again in `ip link add` and friends as I realized that `0` is a valid value for txqueuelen which I previously ruled out by mistake. Thus, I started again and created a customized config parser where I chose `UINT32_MAX` to take the role of the magic value representing the directive being unset while on the other hand disallowing it as a valid input value as explained in the modified commit message. So please take another look - I am looking forward for feedback.

OnkelUlla commented:
> Hmm, instead of .link file, maybe we should add this in .network file?
> Not sure, is the setting is related to the recently added settings by [0a9fb9b](https://github.com/systemd/systemd/commit/0a9fb9bad8b3d245ae4dc27e5fd2675672ddd9ec)?

Ah, I wasn't aware of this commit as I originally developed the feature for a system currently still sitting on _v246.2_.

And yes: I agree that those three parameters should be grouped into one config file (they can be found grouped into the same context in the `ip link` command as well, namely [`ip link add`](https://man7.org/linux/man-pages/man8/ip-link.8.html#SYNOPSIS)). Although I do not have a strong opinion about that I would have expected to find this group in link files as those are only applied once at device birth (and `txqueuelen` perhaps additionally in network files which potentially can be applied more often by analogy to `ip link set`). That's why I implemented it the way I did in the first place. However if you want me to move the mechanics to network files: no problem, I can do that - just let me know. Perhaps the other directives are suited better for link files, too - that could be an outcome as well.

My feeling for link files is that they are responsible for low-level things while network files do the more high-level stuff. Is there a somewhat stricter definition for their field of responsibility?

yuwata commented:
> > Hmm, instead of .link file, maybe we should add this in .network file?
> > Not sure, is the setting is related to the recently added settings by [0a9fb9b](https://github.com/systemd/systemd/commit/0a9fb9bad8b3d245ae4dc27e5fd2675672ddd9ec)?
> 
> My feeling for link files is that they are responsible for low-level things while network files do the more high-level stuff. Is there a somewhat stricter definition for their field of responsibility?

Yeah, that makes sense. After reconsidering that, now I think the settings added by 0a9fb9bad8b3d245ae4dc27e5fd2675672ddd9ec should be moved to .link, instead of introducing `TxQueueLen=` to .network.

@ssahani WDYT?

yuwata commented on src/libsystemd/sd-netlink/netlink-util.c:
```diff
@@ -58,14 +58,15 @@ int rtnl_set_link_name(sd_netlink **rtnl, int ifindex, const char *name) {
         assert(ifindex > 0);
 
-        if (!alias && !mac && mtu == 0)
+        if (!alias && !mac && txqueuelen == (uint32_t) -1 && mtu == 0)
```
Please use `UINT32_MAX`.

OnkelUlla replied:
Will do that.

yuwata commented on src/libsystemd/sd-netlink/netlink-util.c:
```diff
@@ -90,6 +91,12 @@ int rtnl_set_link_properties(sd_netlink **rtnl, int ifindex, const char *alias,
                         return r;
         }
 
+        if (txqueuelen < (uint32_t) -1) {
```
Same here.

OnkelUlla replied:
Yes.

yuwata commented on src/udev/net/link-config.c:
```diff
@@ -150,6 +150,7 @@ int link_load_one(link_config_ctx *ctx, const char *filename) {
                 .rx_flow_control = -1,
                 .tx_flow_control = -1,
                 .autoneg_flow_control = -1,
+                .txqueuelen = (uint32_t) -1,
```
Also here.

OnkelUlla replied:
Okay.

yuwata commented on src/udev/net/link-config.c:
```diff
@@ -701,6 +702,40 @@ int config_parse_ifalias(
+        uint32_t k, *v = data;
+        int r;
+
+        *v = (uint32_t) -1;
```
Please drop this. If an invalid string is assigned, then we should not change anything.

OnkelUlla replied:
Will be dropped.

yuwata commented on src/udev/net/link-config.c:
```diff
@@ -701,6 +702,40 @@ int config_parse_ifalias(
+
+        *v = (uint32_t) -1;
+
+        if (isempty(rvalue))
+                return 0;
```
Then,
```
if (isempty(rvalue)) {
    *v = UINT32_MAX;
    return 0;
}
```

OnkelUlla replied:
Result of the change above.

yuwata commented on src/udev/net/link-config.c:
```diff
@@ -701,6 +702,40 @@ int config_parse_ifalias(
+                log_syntax(unit, LOG_WARNING, filename, line, r, "Failed to parse %s=, ignoring assignment: %s.", lvalue, rvalue);
+                return 0;
+        }
+        if (k == (uint32_t) -1) {
```
`UINT32_MAX`

OnkelUlla replied:
Last occurence.

ssahani commented:
Ok lets move this and reduce .link from .network file

poettering commented:
one more thing: not a fan of needless abbreviations. i.e. I'd call this TxQueueLength=

poettering commented:
> My feeling for link files is that they are responsible for low-level things while network files do the more high-level stuff. Is there a somewhat stricter definition for their field of responsibility?

well, it's primarily about whether something is a property of a device, or a property of the network that is connected to that device. the line is blurry though, as various things that are a local hw property (mtu for example) also have an effect on the outside (as suddenly much larger packets might show up on the wire if you fiddle with the mtu). So, queue length sounds like a mostly internal implementation detail of the linux ip stack and should not be directly visible from the outside, hence I'd say it belongs into .link.

(that is, until someone shows up and can make a really good case that they want to switch the tx queue length depending on the network they connect to, but i#d wait until that happens, and then have the setting in both [Link] sections, in .link and in .network)

OnkelUlla commented:
> one more thing: not a fan of needless abbreviations. i.e. I'd call this TxQueueLength=

I am going to force-push a new version in a second where I changed that.

OnkelUlla commented:
> > My feeling for link files is that they are responsible for low-level things while network files do the more high-level stuff. Is there a somewhat stricter definition for their field of responsibility?
> 
> well, it's primarily about whether something is a property of a device, or a property of the network that is connected to that device. the line is blurry though, as various things that are a local hw property (mtu for example) also have an effect on the outside (as suddenly much larger packets might show up on the wire if you fiddle with the mtu). So, queue length sounds like a mostly internal implementation detail of the linux ip stack and should not be directly visible from the outside, hence I'd say it belongs into .link.
> 
> (that is, until someone shows up and can make a really good case that they want to switch the tx queue length depending on the network they connect to, but i#d wait until that happens, and then have the setting in both [Link] sections, in .link and in .network)

Thanks for taking the time to explain!

yuwata commented:
LGTM. But sorry, recently merged PR conflicts with this. Could you rebase this PR?

OnkelUlla commented:
> LGTM. But sorry, recently merged PR conflicts with this. Could you rebase this PR?

No problem, fixed the conflicts and added a cleanup commit that supplements a log message which has been missed in #17628.

BTW: thanks for guiding me well through the merge of my pull requests!

ssahani commented on man/systemd.link.xml:
```diff
@@ -409,6 +409,13 @@
+        <listitem>
+          <para>Specifies the transmit queue length of the device in number of packets. An unsigned integer
+          in the range 0..4294967294. When unset, the kernel's default will be used.</para>
+        </listitem>
```
https://elixir.bootlin.com/linux/v4.2/source/net/core/dev.c#L6911  range is 1..4294967294

OnkelUlla replied:
Thanks for reviewing, ssahani!
However I think that you are not correct: the kernel's function [`alloc_netdev_mqs()`](https://elixir.bootlin.com/linux/v5.11-rc3/source/net/core/dev.c#L10497) you have found is the one that is used to set the number of transmit/receive queues e.g. as a result of [having received an apropriate netlink message](https://elixir.bootlin.com/linux/v5.11-rc3/source/net/core/rtnetlink.c#L3141) (for example by utilizing the mechanics that you implemented in PR #17635). Sending a netlink message to change the transmit queue length ends in the kernel's function [`do_setlink()`](https://elixir.bootlin.com/linux/v5.11-rc3/source/net/core/rtnetlink.c#L2728) which itself calls [`dev_change_tx_queue_len()`](https://elixir.bootlin.com/linux/v5.11-rc3/source/net/core/dev.c#L8648). And there it is only checked for the range of an unsigned integer which is 0..4294967295.

ssahani commented on src/udev/net/link-config.c:
```diff
@@ -701,6 +704,40 @@ int config_parse_ifalias(
+                log_syntax(unit, LOG_WARNING, filename, line, r, "Failed to parse %s=, ignoring assignment: %s.", lvalue, rvalue);
+                return 0;
+        }
+        if (k == UINT32_MAX) {
```
'0' check required.

OnkelUlla replied:
As discussed above the code should be okay as is.
